### ISSUE TITLE
[Bug] Reused cache_loc host buffer can be overwritten while an in-flight batch still reads it (nondeterministic cross-request corruption)

### ISSUE THREAD

**Issue by real3stone** [sgl-project/sglang-jax#1503]
### Describe the bug

`ScheduleBatch._merge_cache_loc` returns a **view** into the persistent, reused
`cache_loc_host_buf` (allocated once in `ReqToTokenPool.init_cache_loc_host_buffer`).
The comment at the return site says:

    cache_loc_cpu is a view into the reusable host_buf; PD eager-stash
    can overwrite it via _disp(nxt) before this batch's H2D consumes the
    view. Single-threaded (native/colocated) callers don't need the copy.

The single-threaded assumption does not seem to hold. JAX dispatch is
asynchronous, so the scheduler thread can start the next step's
`_merge_cache_loc`, overwriting the buffer in place, while the previous batch's
forward still consumes the view. On the CPU backend, where `device_put` of a
numpy array can be zero-copy, the "device" array may alias the host buffer for
its whole lifetime, so the overwrite lands in the `cache_loc` an in-flight
computation is reading. The result: with concurrent requests, outputs turn into
nondeterministic garbage that leaks content across requests (request B's
completion containing request A's subject matter).

Setup: CPU serving, native backend, `page_size=4`, greedy sampling, 3 requests
arriving 4s apart, so each prefill is single-sequence and the decodes batch
to 2-3. Observations:

* The corruption is nondeterministic: identical inputs and server state produce
  different corruption patterns across runs (all three requests garbled with
  cross-request content in one run, a single request affected in another).
* Ruled out: the decode-mask page-alignment bug (ran with the fix from #1502
  applied, and instrumented `forward_attention` with `jax.debug.print` in the
  live server; the captured batch>1 decode `cache_loc` block layout matches
  `cumsum(aligned_lens)` exactly, so the mask windows land on the right slots);
  mixed chunked prefill (`enable_mixed_chunk` is off); radix cache reuse
  (reproduced after `/flush_cache`).
* Changing the non-pathways return path to an unconditional
  `return cache_loc_cpu.copy()` (the escape hatch the pathways branch already
  uses) made the previously-corrupted requests produce stable, clean output
  across 3/3 repeated runs.
* One residual anomaly is not explained by the copy: the request whose prefill
  runs at the deepest decode concurrency still shows intermittent wrong first
  tokens (2/3 runs, no cross-request content). Possibly a second, related
  host-state hazard on the extend path; the copy does not fix everything.

The aliasing story is a working hypothesis for why the copy helps. What the
experiments show directly: the corruption tracks the buffer reuse and
disappears for the affected requests once the view is copied. I have only
tested the CPU backend; whether the TPU H2D path can read a partially
overwritten buffer depends on when the transfer touches host memory. This
failure mode produces the same concurrency-dependent garbage reported in
#1297, so it may be a second contributing factor there beyond the mask bug.

Possible fixes: (a) an unconditional `.copy()` (the buffer is a small int32
array; validated above); (b) double-buffering the host buffer; (c)
synchronizing on the previous step's transfer before rewriting. Happy to send a
PR for (a) or (b) with a regression test if the diagnosis is confirmed.

### Reproduction

Server (CPU):

    python -m sgl_jax.launch_server --model-path Qwen/Qwen3-0.6B \
      --device cpu --attention-backend native --page-size 4 \
      --max-running-requests 4 --context-length 1024 --max-total-tokens 4096 \
      --max-prefill-tokens 1024 --dtype bfloat16 --skip-server-warmup \
      --disable-precompile --port 30000

Client: three `/generate` requests with `temperature=0`, `max_new_tokens=48`,
started 4 seconds apart (threads), so each prefill is a single-sequence batch
and the decodes overlap into a batch of 2-3. Compare against the same three
prompts run one at a time. Corruption appears within the first few decode
tokens of the overlap window; repeating the run changes which requests are
affected. I can attach the client script and the captured solo/staggered
outputs (with and without the `.copy()` patch).

### Environment

    Python: 3.12.13 (main, Jun 24 2026, 04:58:23) [GCC 14.2.0]
    sglang-jax: main @ d8b61193 (+ the #1297 decode-mask fix from #1502, needed to
                rule the mask out as the cause; the buffer-reuse code itself is
                unmodified main)
    jax: 0.10.2
    jaxlib: 0.10.2
    transformers: 4.57.6
    numpy: 2.2.6
    device: [cpu-0]: cpu:0
